### ISSUE TITLE
[TechDocs] Add an in-context search bar to an individual TechDocs site

### ISSUE THREAD

**Issue by iamEAP** [backstage/backstage#4781]
While we're still implementing [a more complete search platform](https://github.com/backstage/backstage/projects/6) for Backstage generally, it's possible to implement a simple, temporary "search this TechDocs site" experience on an individual TechDocs site.

## Feature Suggestion

Add a search bar above the contents/nav of a TechDocs site that can be used to search for pages/paragraphs within the TechDocs site that, when selected, navigates you to the page/paragraph.

An extremely high fidelity 😬  mock-up of that might look like...

![Search Bar](https://user-images.githubusercontent.com/3496491/109808334-4352ca00-7c27-11eb-9e74-162fa23f0df6.png)

![Search Autocomplete Results](https://user-images.githubusercontent.com/3496491/109808347-477ee780-7c27-11eb-80d8-b1da6a678d68.png)

## Possible Implementation
(Once [Search V1 Milestone](https://github.com/backstage/backstage/milestone/26) is Complete)
- Leverage existing, reusable Search components and context from the Search plugin to build the search experience into the `<Reader />` component
- Pre-filter the search context to just TechDocs documents belonging to this entity.

#### Alternative Implementation
- Consider using an off-the-shelf [autocomplete component](https://material-ui.com/components/autocomplete/#asynchronous-requests) so that search results can be shown and selected on-page.
- Consider using [lunr](https://lunrjs.com/docs/lunr.html) to create a client-side, in-process index and query it based on the input
- An individual site's lunr index (represented as JSON) can be loaded from a location like [/api/techdocs/static/docs/default/Component/backstage/search/search_index.json](https://demo.backstage.io/api/techdocs/static/docs/default/Component/backstage/search/search_index.json)

**Comment by guillermomanzo** [comment]
hey @iamEAP We, at Expedia Group are looking to take this on.

**Comment by rodmachen** [comment]
Yes, as Guillermo said, the team here at Expedia been looking at this issue and would like to contribute this feature to the Backstage project. 

@iamEAP please confirm and let us know anything else that would be helpful. The team at Expedia has contributed other items before and look forward to doing so again. We'll have our designers and developers start looking at implementation for this feature.

Thank you.

**Comment by iamEAP** [comment]
Thanks @rodmachen and @guillermomanzo!  That's great news.  Hopefully the `Possible Implementation` section above provides enough of a direction for y'all to get started.  Otherwise, feel free to reach out here or Discord.

For now, I'll assign to both of y'all, just to indicate the issue is being worked on.

**Comment by rodmachen** [comment]
Just now getting this prioritized in our backlog. Should be able to start working on it soon @iamEAP .

**Comment by iamEAP** [comment]
Glad to hear it!  Let us know if we can assist in any way.

**Comment by iamEAP** [comment]
Hey @rodmachen, just a quick heads up related to ongoing Backstage search platform improvements.  We're making pretty steady progress on a more robust search backend that _could_ be used to power this in-context search experience (rather than building a custom autocomplete experience on top of an individual site's `search_index.json` file).

It's not 100% there yet, but we're getting close to it being possible.  And while it's not in our [current milestone](https://github.com/backstage/backstage/milestone/26) to add TechDocs support, by the end of the milestone, the _ability_ to do so would be there.  Which would mean there'd be opportunity to implement this feature on top of the Search Platform, rather than through the `search_index.json` file directly, as outlined in the issue description.

With that in mind, wanted to check in on where y'all are at on this.  How's it going?

**Comment by adrianke77** [comment]
Hi @rodmachen, would like to gently inquire is work is ongoing for this issue? As we're looking forward to it at our side (Grab)

**Comment by rodmachen** [comment]
@adrianke77 #4568 needs to be done first, and we're in conversations about that work.

**Comment by yousifalraheem** [comment]
Hi, is there any progress being made for this issue? We at Zalando are really interested in this feature. In fact, it's in the top 3 of the highest requested features. We are happy to contribute if necessary.

I found this video on YouTube were the search is implemented in the internal version of Backstage.
[![How to make great documentation](https://img.youtube.com/vi/mOLCgdPw1iA/0.jpg)](https://www.youtube.com/watch?v=mOLCgdPw1iA?t=107)

Can we just borrow some of the implementation for the open source version?

**Comment by Xantier** [comment]
Hi @yousifalraheem,

The implementation of this has started on the backend side which introduces a collator to gather the actual data to be indexed for the techdocs search (https://github.com/backstage/backstage/pull/6375). That is advancing nicely and I like to think fairly close to being merged in (I'll probably add one more test case for it) and would then unblock this ticket to being started. 

I will likely have time to work on this in the coming weeks unless someone nabs it first.

**Comment by laceyallyn** [comment]
Hey there, I'm a the Product Manager for Backstage in Zalando, and I'd really like and update on this topic, as it's a topic that's coming up over and over from our users. We already have a global search across documentation and it's just not solving the problems our users need it to, and it's highly impacting adoption at the moment.

Without a solid timeline for this, we are seriously considering moving away from the open source tech docs plugin, and I'd rather it not come to that.

**Comment by Fox32** [comment]
@laceyallyn With the techdocs collator (#6375) and the search v1 milestone now it place, this is a frontend only change. You can query the search backend and scope it to a single entity by passing filter queries (`name`, `kind`, `namespace`), filter it by `type` `techdocs` and pass the user supplied `term`. I guess the main issue is finding a way to visualize the results.

The issue is marked as help wanted, does your team want to jump in?
